src/main/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiController.java

```diff
@@ -145,8 +149,60 @@ public ControlledGcpResourceApiController(
     final AuthenticatedUserRequest userRequest = getAuthenticatedInfo();
     ControlledResource controlledResource =
         controlledResourceService.getControlledResource(workspaceId, resourceId, userRequest);
-    ApiGcpGcsBucketResource response = controlledResource.castToGcsBucketResource().toApiResource();
-    return new ResponseEntity<>(response, HttpStatus.OK);
+    try {
+      ApiGcpGcsBucketResource response =
+          controlledResource.castToGcsBucketResource().toApiResource();
+      return new ResponseEntity<>(response, HttpStatus.OK);
+    } catch (InvalidMetadataException ex) {
+      throw new BadRequestException(
+          String.format(
+              "Resource %s in workspace %s is not a controlled GCS bucket.",
+              resourceId, workspaceId));
+    }
+  }
+
+  @Override
+  public ResponseEntity<ApiGcpBigQueryDatasetResource> getBigQueryDataset(
+      UUID workspaceId, UUID resourceId) {
+    final AuthenticatedUserRequest userRequest = getAuthenticatedInfo();
+    String projectId = workspaceService.getRequiredGcpProject(workspaceId);
+    ControlledResource controlledResource =
+        controlledResourceService.getControlledResource(workspaceId, resourceId, userRequest);
+    try {
+      ApiGcpBigQueryDatasetResource response =
+          controlledResource.castToBigQueryDatasetResource().toApiResource(projectId);
```

Comment:
I think this casting approach is not working so well. I have a thought. What if each controlled resource subclass provided two methods. For example:
```
public static from(ControlledResource resource)
```
and
```
public static fromRequest(ControlledResource resource)
```
We would use `fromRequest()` when the inputs were from the user. It would throw BadRequestException if it isn't the right type. The `from()` we would use internally when we are sanity checking our own logic. That can throw InvalidMetadataException.

This idea is probably not for this PR, but I think it would avoid putting stuff in ControlledResource.java per-sub-class, so reduce the touch points when adding a new controlled resource. And it would remove all of the try/throw-something-different copies.

Reply:
SGTM, I'm not a fan of catching unchecked exceptions here and we could avoid that by separating "cast and fail with 400" from "cast and fail with 500" into separate methods. We'd still have a type-specific casting method, but at least it would live inside the specific resource class, which seems better.

Reply:
Or parameterize the cast method with the type you want thrown.